src/components/menu/menu-popover.tsx

```diff
@@ -1,29 +1,70 @@
-import { type FC, type MutableRefObject, useRef } from 'react'
+import { type FC, type MutableRefObject, useLayoutEffect, useRef, useState } from 'react'
 import { useClickOutside } from '../../hooks/use-click-outside'
 import { useMenuContext } from './menu-context'
 import { ElMenuPopover } from './styles'
 
-export const MenuPopover: FC = ({ children }) => {
+export const calculatePopoverPosition = (
+  container: HTMLElement,
+  popover: HTMLDivElement,
+  setPopoverStyle: React.Dispatch<React.SetStateAction<React.CSSProperties>>,
+  additionalGap = 0,
+) => {
+  const triggerBtn = container.querySelector('[role="button"]')
+  if (triggerBtn) {
+    const viewportHeight = window.innerHeight
+    const popoverHeight = popover.getBoundingClientRect().height
+    const { bottom: triggerBottomPos, height: triggerheight } = triggerBtn.getBoundingClientRect()
+    const gap = 3 // the shadow width or .etc
+    const spaceBelowButton = viewportHeight - triggerBottomPos
+
+    const top = popoverHeight > spaceBelowButton ? 0 - popoverHeight - gap : triggerheight + gap
+
+    setPopoverStyle({ top: top + additionalGap })
+  }
+}
+
+export const MenuPopover: FC<{
+  /**
+   * Optional parameter to adjust the vertical position of the popover
+   *
+   * @default 0
+   */
+  additionalGap?: number
+}> = ({ children, additionalGap }) => {
   const { isOpen, closeMenu, getPopoverProps } = useMenuContext()
   const popoverRef = useRef<HTMLDivElement>(null)
 
+  const [popoverStyle, setPopoverStyle] = useState<React.CSSProperties>({
+    position: 'absolute',
+  })
+
   useClickOutside(popoverRef, closeMenu)
 
-  if (isOpen) {
-    const onClick = (event) => {
-      // close menu if clicked target has parent with below selector
-      const menuItem = event.target.closest('[data-close-menu="true"]')
-      if (menuItem) closeMenu()
+  useLayoutEffect(() => {
+    const container = popoverRef.current?.parentElement
+    if (container && isOpen && popoverRef.current) {
+      calculatePopoverPosition(container, popoverRef.current, setPopoverStyle, additionalGap)
     }
+  }, [isOpen])
+
+  if (!isOpen) return null
 
-    return (
-      <ElMenuPopover ref={popoverRef as MutableRefObject<HTMLDivElement>} {...getPopoverProps()} onClick={onClick}>
-        {children}
-      </ElMenuPopover>
-    )
+  const handleClick = (event) => {
+    // close menu if clicked target has parent with below selector
+    const menuItem = event.target.closest('[data-close-menu="true"]')
+    if (menuItem) closeMenu()
   }
 
-  return null
+  return (
+    <ElMenuPopover
+      style={popoverStyle}
+      ref={popoverRef as MutableRefObject<HTMLDivElement>}
```

Comment:
**question:** I don't understand why you're needing to use a type assertion for the ref here?

Reply:
something wrong with my linter but its good now, turns out it doesn't need to have that assertion👍